messybrainz/create_recording_clusters.py

```diff
@@ -0,0 +1,208 @@
+
+    Returns:
+        List of gids.
+    """
+
+    query = text("""SELECT gid
+                      FROM recording_json
+                      JOIN recording
+                        ON recording_json.id = recording.data
+                     WHERE recording_json.data ->> 'recording_mbid' = :recording_mbid
+    """)
+    gids = connection.execute(query, {
+        "recording_mbid": recording_mbid,
+        }
+    )
+    return [gid[0] for gid in gids]
+
+
+def fetch_distinct_recording_mbids(connection):
+    """Fetch all the distinct recording MBIDs we have in recording_json table.
+
+    Args:
+        connection: the sqlalchemy db connection to be used to execute queries
+
+    Returns:
+        recording_mbids.
+    """
+
+    query = text("""SELECT DISTINCT data ->> 'recording_mbid' AS recording_mbids
+                               FROM recording_json 
+                              WHERE data ->> 'recording_mbid' IS NOT NULL
+                """)
+    recording_mbids = connection.execute(query)
+    return recording_mbids
+
+
+def link_recording_mbid_to_recording_msid(connection, cluster_id, mbid):
+    """Links the recording mbid to the cluster_id.
+
+    Args:
+        connection: the sqlalchemy db connection to be used to execute queries
+        cluster_id: the gid which represents the cluster.
+        mbid: mbid for the cluster.
+
+    Returns:
+        True if link is inserted into the cluster otherwise False is returned.
+    """
+
+    cluster_added = False
+    if not is_recording_cluster_present_in_recording_redirect(connection, cluster_id, mbid):
+        query = text("""INSERT INTO recording_redirect (recording_cluster_id, recording_mbid)
+                            VALUES (:cluster_id, :mbid)
+                    """)
+        connection.execute(query, {
+            "cluster_id": cluster_id,
+            "mbid": mbid,
+        })
+        cluster_added = True
+    return cluster_added
+
+
+def is_recording_cluster_present_in_recording_redirect(connection, cluster_id, mbid):
+    """Checks if recording cluster is already present in the recording_redirect table.
+
+    Args:
+        connection: the sqlalchemy db connection to be used to execute queries
+        cluster_id: the gid which represents the cluster.
+        mbid: mbid for the cluster.
+
+    Returns:
+        True if cluster already present in recording_redirect table otherwise 
+        False is returned.
+    """
+
+    query = text("""SELECT recording_cluster_id, recording_mbid
+                      FROM recording_redirect
+                     WHERE recording_cluster_id = :recording_cluster_id
+                       AND recording_mbid = :recording_mbid
+    """)
+    result = connection.execute(query, {
+        "recording_cluster_id": cluster_id,
+        "recording_mbid": mbid,
+    })
+    if result.rowcount:
+        return True
+    return False
+
+
+def is_recording_cluster_present_in_recording_cluster(connection, cluster_id, gid):
```

Comment:
Ideally, both of these functions should not be needed if we add the right table contraints as mentioned above. Lets see if we can get rid of these functions, shall we?